### PR TITLE
Core: Optimize manifest evaluation for super wide tables

### PR DESCRIPTION
**Pull request by irshadcc**
During the snapshot commit process of MergingSnapshotProducer, Iceberg tries to merge the manifest files to increase the planning performance. During operations like overwrite, MergingSnapshotProducer finds the matching manifest files by deleteExpression and rewrites the manifests by filtering out the deleted manifest entries.

https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/ManifestFilterManager.java#L330

While finding the matching manifests by deleteExpression, we found that the Schema is created every time the expression needs to be bound. This has proven very expensive when there are large number of manifest files (~25,000 manifest files) for a super wide table (~35,000 columns).

https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/expressions/NamedReference.java#L41

For optimising the manifest evaluation, we can add a method called asSchema() in the StructType class to avoid creating the new Schema every time the filter needs to be bound.

### CODE REVIEW

**Comment by irshadcc** [comment]
Resolves issue #9118

File: api/src/main/java/org/apache/iceberg/types/Types.java [hunk 1]
```diff
@@ -592,6 +594,13 @@ public Types.StructType asStructType() {
       return this;
     }
 
+    public Schema asSchema() {
```
Comment (Fokko) [inline api/src/main/java/org/apache/iceberg/types/Types.java:597]: Could you add a Javadoc here to indicate that it is for optimization

**Review by Fokko — APPROVED** [review]
Thanks for raising this @irshadcc, this looks good to me. I've left two small comments, could you take a peek at those? Thanks for fixing this! 🙌

**Comment by irshadcc** [comment]
> Thanks for raising this @irshadcc, this looks good to me. I've left two small comments, could you take a peek at those? Thanks for fixing this! 🙌

I've added the Javadoc and removed the empty.

**Comment by irshadcc** [comment]
@Fokko Can we merge this PR ?

**Comment by irshadcc** [comment]
kind ping @Fokko

**Comment by Fokko** [comment]
Hey @irshadcc Sorry for the long wait here, and thanks for pinging me. Let's get this in 🚀

**Pull request merged**